torch/distributions/utils.py

```diff
@@ -130,19 +131,34 @@ def probs_to_logits(probs, is_binary=False):
     return torch.log(ps_clamped)
 
 
-class lazy_property:
+T = TypeVar("T", covariant=True)
+
+
+class lazy_property(Generic[T]):
     r"""
     Used as a decorator for lazy loading of class attributes. This uses a
     non-data descriptor that calls the wrapped method to compute the property on
     first call; thereafter replacing the wrapped method into an instance
     attribute.
     """
 
-    def __init__(self, wrapped):
-        self.wrapped = wrapped
+    def __init__(self, wrapped: Callable[..., T]) -> None:
```

Comment:
```suggestion
    def __init__(self, wrapped: Callable[..., T]) -> None:

Reply:
I don't think a `ParamSpec` makes too much sense here, if one wants more precise types one could make it a `Callable[[Any], T]` or `Callable[[S], T]`. But explicitly allowing arbitrary signatures makes no sense for a property to me. (in contrast, `Callable[..., T]` is a gradual type, i.e. it says that the concretely allowed signatures are not specified). For comparison, here is how Eric Traut once suggested annotating `property`: https://github.com/python/typing/issues/985#issuecomment-1002770338

Reply:
Instead of `ParamSpec`, I added a second type var for the object type. Methods wrapping properties generally should only have one argument and no optional arguments. The only place I see where this could be problematic is if you want to make a function a property that originates from `functools.partial`.

Reply:
I am a bit confused, does TypeVar require callable args to be kw_only? TypeVarTuple does for sure, but I don't think that's what want here... I am worried the TypeVar isn't being expanded how you want

Reply:
Yeah, it looks like to me TypeVar does not enforce that the input to the callable is position only. Better to either parameterize it in ParamSpec (and maybe add some typing asserts). Callable itself does not have a concept of pos args vs kwargs it seems, they are mashed together. You can try to expand a TypeVarTuple to get something closer to what you want, but I am not sure that would be valid either.

Reply:
I do not see how or why this is a problem in this case. FWIW, `self` by itself is not positional-only to begin with (though type checkers treat it that way, but that is an implementation detail of the type checker).

Reply:
Can you give an example of a bad case you want to avoid?

Reply:
Oh then why not just ParamSpec then? It will capture just the positional only argument self?